plugins/filter/keep_keys.py

```diff
@@ -0,0 +1,154 @@
+
+  # The results of all examples are all the same.
+  r:
+    - {k0_x0: A0, k1_x1: B0}
+    - {k0_x0: A1, k1_x1: B1}
+'''
+
+RETURN = '''
+  _value:
+    description: The list of dictionaries with selected keys.
+    type: list
+    elements: dictionary
+'''
+
+from ansible.errors import AnsibleFilterError
+from ansible.module_utils.six import string_types
+from ansible.module_utils.common._collections_compat import Mapping, Sequence
+
+import re
+
+
+def keep_keys(data, target=None, matching_parameter='equal'):
+    """keep specific keys from dictionaries in a list"""
+
+    ld = data
+    tt = target
+    mp = matching_parameter
+    ml = ['equal', 'starts_with', 'ends_with', 'regex']
+
+    if not isinstance(ld, Sequence):
+        msg = "First argument for keep_keys must be a list. %s is %s"
+        raise AnsibleFilterError(msg % (ld, type(ld)))
+
+    for elem in ld:
+        if not isinstance(elem, Mapping):
+            msg = "Elements of the data list for keep_keys must be dictionaries. %s is %s"
+            raise AnsibleFilterError(msg % (elem, type(elem)))
+
+    if mp not in ml:
+        msg = ("The matching_parameter for keep_keys must be one of %s. matching_parameter is %s")
+        raise AnsibleFilterError(msg % (ml, mp))
+
+    if mp == 'regex':
+        if isinstance(target, string_types):
+            tt = target
+        elif isinstance(target, Sequence):
+            tt = target[0]
+        else:
+            msg = ("The target for keep_keys must be a string or a list if matching_parameter is regex."
+                   "target is %s.")
+            raise AnsibleFilterError(msg % target)
+        try:
+            re.compile(tt)
+            is_valid = True
+        except re.error:
+            is_valid = False
+        if not is_valid:
+            msg = ("The target for keep_keys must be a valid regex if matching_parameter is regex."
+                   "target is %s")
+            raise AnsibleFilterError(msg % tt)
+    else:
+        if not isinstance(tt, Sequence):
+            msg = ("The target for keep_keys must be a list if matching_parameter is %s. %s is %s")
+            raise AnsibleFilterError(msg % (mp, tt, type(tt)))
+        for elem in tt:
+            if not isinstance(elem, string_types):
+                msg = "Elements of the targets for keep_keys must be strings. %s is %s"
+                raise AnsibleFilterError(msg % (elem, type(elem)))
+
+    if mp == 'equal':
+        my_keys = [[k for k in i.keys() if k in tt] for i in ld]
+    elif mp == 'starts_with':
+        my_keys = [[k for k in i.keys() if k.startswith(tuple(tt))] for i in ld]
+    elif mp == 'ends_with':
+        my_keys = [[k for k in i.keys() if k.endswith(tuple(tt))] for i in ld]
+    elif mp == 'regex':
+        if isinstance(target, string_types):
+            tt = target
+        else:
+            tt = target[0]
+        my_keys = [[k for k in i.keys() if re.match(tt, k)] for i in ld]
```

Comment:
In case of `mp == 'regex'`, you already compiled the regular expression. But now you're using `re.match()` to match the regular expression *as a string* against a large set of strings. You should really use the compilation result here.

(The current version is likely still efficient because `re` does some magic to cache recently used regular expressions, but simply relying on that is not a good idea, and also still less efficient than simply using the compiled version you already could have here.)